src/components/common/dialog.tsx

```diff
@@ -0,0 +1,1550 @@
+      ...
+        <DialogTextInput value={value.sub("first")} />
+      ...
+    }
+
+  It would be used in a dialog like this:
+
+    interface DialogValues {
+      person: PersonValue;
+    }
+
+    function init() {
+      return {
+        person: init_Person(...)
+      }
+    }
+
+    function validate() {
+      validate_Person(dlg.value("person"));
+    }
+
+    const dlg = useDialogState(init, validate);
+
+    return (
+      ...
+      <Person value={dlg.value("person")} />
+      ...
+    );
+
+  Here is a pattern for handling types that include alternatives, such
+  as "PersonValue | string".  This could be used to encode either the
+  state for a working Person component, or an excuse message that
+  explains why it can't work.
+
+    function init_Person(...): PersonValue | string {
+      if (some condition)
+        return _("Something wrong.");
+
+      return { ... };
+    }
+
+    function validate_Person(value: DialogValue<PersonValue | string>) {
+      const val = value.get();
+      if (typeof val == "string")
+          return;
+
+      const p_value = value.at(val);
+
+      p_value.sub("first").validate(v => {
+        if (!v)
+          return "First name can't be empty";
+      })
+    }
+
+    export const Person = ({ value } : { value: DialogValue<PersonValue | string> }) => {
+      const val = value.get();
+      if (typeof val == "string")
+          return null;
+
+      const p_value = value.at(val);
+
+      const { country, first, last } = p_value.get();
+      ...
+        <DialogTextInput value={p_value.sub("first")} />
+      ...
+    }
+
+  Note the use of the "value.at()" function to get a handle for a
+  PersonValue that can be used to access the "first" sub value, etc.
+
+ */
+
+import React, { useState } from "react";
+import { useObject, useInit, useOn } from 'hooks';
+import { EventEmitter } from 'cockpit/event';
+
+import cockpit from "cockpit";
+
+import { Button, type ButtonProps } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { FormGroup, type FormGroupProps, FormHelperText } from "@patternfly/react-core/dist/esm/components/Form";
+import { TextInput, type TextInputProps } from "@patternfly/react-core/dist/esm/components/TextInput";
+import { Alert } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import {
+    HelperText, HelperTextItem, type HelperTextItemProps
+} from "@patternfly/react-core/dist/esm/components/HelperText";
+import { Checkbox } from "@patternfly/react-core/dist/esm/components/Checkbox";
+import {
+    FormSelect, FormSelectOption, type FormSelectProps,
+} from "@patternfly/react-core/dist/esm/components/FormSelect";
+import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
+
+const _ = cockpit.gettext;
+
+function debug(...args: unknown[]) {
+    if (window.debugging == "all" || window.debugging?.includes("dialog"))
+        console.debug("dialog:", ...args);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-4e45c3b5-20251217-171357-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)